tests/unittests/ExecutorTest.cpp

```diff
@@ -0,0 +1,501 @@
+  auto deviceManager = std::make_shared<TestDeviceManager>();
+  deviceManagerMap_.insert(std::make_pair(testDeviceId, deviceManager));
+
+  // Create and allocate all Placeholders and initialize the Tensors.
+  auto testType = new Type(ElemKind::FloatTy, {1, 2, 2});
+  auto alphaIn = std::make_unique<Placeholder>("alphaIn", testType, false);
+  ctx_.allocate(alphaIn.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto alphaOut = std::make_unique<Placeholder>("alphaOut", testType, false);
+  ctx_.allocate(alphaOut.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto betaIn = std::make_unique<Placeholder>("betaIn", testType, false);
+  ctx_.allocate(betaIn.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto betaOut = std::make_unique<Placeholder>("betaOut", testType, false);
+  ctx_.allocate(betaOut.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto deltaIn = std::make_unique<Placeholder>("deltaIn", testType, false);
+  ctx_.allocate(deltaIn.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto deltaOut = std::make_unique<Placeholder>("deltaOut", testType, false);
+  ctx_.allocate(deltaOut.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto epsIn = std::make_unique<Placeholder>("epsIn", testType, false);
+  ctx_.allocate(epsIn.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+  auto epsOut = std::make_unique<Placeholder>("epsOut", testType, false);
+  ctx_.allocate(epsOut.get())->init(Tensor::InitKind::Xavier, 1.0, rng_);
+
+  // Build the DAG. The DAG created below looks like this:
+  //           root
+  //         /      \
+  //        v       v
+  //      alpha    beta
+  //        \       /
+  //         v     v
+  //          gamma
+  //         /    \
+  //        v     v
+  //     delta   eps
+  //
+  ExecutorTestDAG dag;
+  DAGNode *alpha = dag.addNode("alpha", testDeviceId, /*parents=*/{},
+                               /*inputs=*/{alphaIn.get()},
+                               /*outputs=*/{alphaOut.get()});
+  DAGNode *beta = dag.addNode("beta", testDeviceId, /*parents=*/{},
+                              /*inputs=*/{betaIn.get()},
+                              /*outputs=*/{betaOut.get()});
+  DAGNode *gamma = dag.addNode("gamma", testDeviceId, /*parents=*/{alpha, beta},
+                               /*inputs=*/{alphaOut.get(), betaOut.get()},
+                               /*outputs=*/{deltaIn.get(), epsIn.get()});
+  dag.addNode("delta", testDeviceId, /*parents=*/{gamma},
+              /*inputs=*/{deltaIn.get()},
+              /*outputs=*/{deltaOut.get()});
+  dag.addNode("eps", testDeviceId, /*parents=*/{gamma},
+              /*inputs=*/{epsIn.get()},
+              /*outputs=*/{epsOut.get()});
+
+  // Create the input Context to pass to Executor::run().
+  auto inputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{alphaIn.get(), betaIn.get()}));
+
+  // Create input/output Contexts for all DAGNodes. The input Contexts are
```

Comment:
The code here is super verbose. I thought about folding it into `ExecutorTestDAG::addNode`, but I think that `ExecutorTestDAG` should only create the DAG so that it can be reused in other places if needed.